File: solution/930. Binary Subarrays With Sum/930_01.cpp
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        unordered_map<int, int> m;
        m[0] = 1;
        int res = 0, presum = 0;
        for (int x : nums) {
            presum += x;
            if (m.count(presum - goal))
                res += m[presum - goal];
            m[presum]++;
        }
        return res;
    }
};

File: solution/930. Binary Subarrays With Sum/930_01.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        seen = defaultdict(int)
        seen[0] = 1
        res, prefix = 0, 0
        for x in nums:
            prefix += x
            if prefix - goal in seen:
                res += seen[prefix - goal]
            seen[prefix] += 1
        return res
    
